### PR TITLE
More ABNF cleanup; Test with CTS and JPC

### LINKED ISSUE #354: ABNF cleanup 2
Two objectives:

(1) aesthetic cleanup, e.g.:
https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/344#issuecomment-1382692603

(2) more tests, e.g., to be built from query corpora we can find

### PR DESCRIPTION
**Pull request by cabo**
Close #354

### CODE REVIEW

File: scripts/test-abnf-cts.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env ruby
+
+require 'abnftt'
+
+parser = ABNF.from_abnf(File.read("draft-ietf-jsonpath-base.abnf"))
+
+require 'json'
+
+cts = JSON.parse(File.read("../jsonpath-compliance-test-suite/cts.json"))
```
Comment (glyn) [inline scripts/test-abnf-cts.rb:9]: Hmmm. This won't be found on most people's machines. Do we need a README? I know this script is really only for limited usage, but...

File: draft-ietf-jsonpath-base.md [hunk 2]
```diff
@@ -697,7 +697,7 @@ single-quoted       = unescaped /
 
 ESC                 = %x5C                           ; \  backslash
 
-unescaped           = %x20-21 /                      ; s. RFC 8259
+unescaped           = %x20-21 /                      ; see RFC 8259
```
Comment (glyn) [inline draft-ietf-jsonpath-base.md:700]: Oh, that's what "s." means! Thank you.

File: scripts/test-abnf-cts.rb (hunk 1 above)
Comment (cabo) [reply scripts/test-abnf-cts.rb:9]: Yes, we may need a README.  Alternatively, the test script could fetch right from Github.

Comment (cabo) [reply scripts/test-abnf-cts.rb]: Fixed in 2d24063

File: draft-ietf-jsonpath-base.md (hunk 2 above)
Comment (cabo) [reply draft-ietf-jsonpath-base.md:700]: Germanism... Thanks for spotting this.

**Comment by gregsdennis** [comment]
Can you explain how the test scripts work and how to interpret the output, please?

**Comment by cabo** [comment]
> Can you explain how the test scripts work and how to interpret the output, please?

I haven't done the work to make this a real CI script.
You clone the JPC to a directory parallel to where you cloned the I-D repo.

kdrfc draft-ietf-jsonpath-base.md
sh scripts/abnfp.sh
ruby scripts/test-abnf-jpc2.rb > scripts/test-abnf-jpc2.out 2>&1
ruby scripts/test-abnf-jpc.rb > scripts/test-abnf-jpc.out 2>&1
ruby scripts/test-abnf-cts.rb > scripts/test-abnf-cts.out 2>&1
ruby scripts/test-abnf.rb > scripts/test-abnf.out 2>&1

Compare the output with the checked-in output.
(The output contains all parse failures, including the intentional ones.
The word FAIL tells you about unexpected output.)
An output that says "successfully completed" at the end has done exactly that.

**Comment by gregsdennis** [comment]
JPC?

**Comment by gregsdennis** [comment]
Is this something that should be in the same repo as the spec?  Seems like it should be with a test suite.  Or do we plan to remove it before publication?

File: draft-ietf-jsonpath-base.md [hunk 3]
```diff
@@ -1817,16 +1816,14 @@ normal-single-quoted = normal-unescaped /
...
-                           ; \b \f \n \r \t
-                           ; b /         ;  BS backspace U+0008
-                           ; t /         ;  HT horizontal tab U+0009
-                           ; n /         ;  LF line feed U+000A
-                           ; f /         ;  FF form feed U+000C
-                           ; r /         ;  CR carriage return U+000D
-                         "'" /           ;  ' apostrophe U+0027
-                         "\" / ; \ backslash (reverse solidus) U+005C
-                         (%x75 normal-hexchar)
+normal-escapable     = (%x62 / ; b  BS backspace U+0008
```
Comment (glyn) [inline draft-ietf-jsonpath-base.md:1819]: Nit: delete extra space before BS.

**Review by glyn — APPROVED** [review]
Approved with two nits.

**Review by glyn — APPROVED** [review]
Looks plausible. I haven't reviewed or run the Ruby code.

**Pull request merged**